message_ix/core.py

```diff
@@ -26,6 +27,111 @@ def __init__(self, mp, model, scenario=None, version=None, annotation=None,
         super().__init__(mp, model, scenario, version, scheme, annotation,
                          cache)
 
+    @lru_cache()
+    def _year_idx(self, name):
+        """Return (idx_set, idx_name) for 'year'-indexed dims of *name*."""
+        # NB Since item dimensionality is fixed, the result can be cached for
+        #    performance
+        return list(filter(
+            lambda e: e[0] == 'year',
+            zip(self.idx_sets(name), self.idx_names(name))
+        ))
+
+    def _year_as_int(self, name, df):
+        """Convert 'year'-indexed columns of *df* to int dtypes."""
+        year_idx = self._year_idx(name)
+        if len(year_idx):
+            return df.astype({c: 'int' for _, c in year_idx})
+        else:
+            return df
+
+    # Override ixmp methods to convert 'year'-indexed columns to int
+    def equ(self, name, filters=None):
+        """Return equation data.
+
+        Same as :meth:`ixmp.Scenario.equ`, except columns indexed by the
+        |MESSAGEix| set ``year`` are returned with :obj:`int` dtype.
+
+        Parameters
+        ----------
+        name : str
+            Name of the equation.
+        filters : dict (str -> list of str), optional
+            Filters for the dimensions of the equation.
+
+        Returns
+        -------
+        pd.DataFrame
+            Filtered elements of the equation.
+        """
+        return self._year_as_int(name, super().equ(name, filters))
```

Comment:
How much efficient is such a post-processing in comparison of having years dimensions converted to integers upon df creation in ixmp core?

Reply:
To be frank, I don't know. I hope pandas' dtype conversion is fast…

Having it done in ixmp core would require the extension to the Backend API that we discussed [here](https://github.com/iiasa/message_ix/issues/268#issuecomment-557885384).